wandb/sdk/wandb_run.py

```diff
@@ -1456,39 +1457,110 @@ def _datatypes_callback(self, fname: str) -> None:
-                        "Vega2", row[k].user_query(f"Custom Chart Tables/{k}_table")
+
+        def transform(key: List[str], value: Any) -> Any:
+            if isinstance(value, dict):
+                return {k: transform([*key, k], v) for k, v in value.items()}
+            if isinstance(value, Visualize):
+                formatted_key = ".".join(key)
+                config_key = value.get_config_key(formatted_key)
+                config_value = value.get_config_value(formatted_key)
+                value = value._data
+                self._config_callback(val=config_value, key=config_key)
+                return value
+            elif isinstance(value, CustomChart):
+                formatted_key = ".".join(key)
+                chart_keys.add(tuple(key))
+                config_key = value.get_config_key(formatted_key)
+                if value._split_table and len(key) == 1:
+                    config_value = value.get_config_value(
+                        "Vega2",
+                        value.user_query(f"Custom Chart Tables/{formatted_key}_table"),
+                    )
+                    split_table_set.add(tuple(key))
+
+                elif value._split_table and len(key) > 1:
+                    config_value = value.get_config_value(
+                        "Vega2",
+                        value.user_query(
+                            f"{key[0]}.Custom Chart Tables/{key[1]}_table"
+                        ),
                     )
-                    split_table_set.add(k)
+                    split_table_set.add(tuple(key))
+
                 else:
-                    value = row[k].get_config_value(
-                        "Vega2", row[k].user_query(f"{k}_table")
+                    config_value = value.get_config_value(
+                        "Vega2", value.user_query(f"{formatted_key}_table")
                     )
-                row[k] = row[k]._data
-                self._config_callback(val=value, key=key)
-
-        for k in chart_keys:
-            # remove the chart key from the row
-            # TODO: is this really the right move? what if the user logs
-            #     a non-custom chart to this key?
-            if k in split_table_set:
-                row[f"Custom Chart Tables/{k}_table"] = row.pop(k)
+                value = value._data
+                self._config_callback(val=config_value, key=config_key)
+                return value
             else:
-                row[f"{k}_table"] = row.pop(k)
+                return value
+
+        for k, v in row.items():
+            row[k] = transform([k], v)
+
+        for chart_key_tuple in chart_keys:
+            chart_key_list = list(chart_key_tuple)
+            value = self._get_nested_value(row, chart_key_list)
+
+            self._set_nested_value(
+                row,
+                chart_key_list,
+                value,
+                split_table=chart_key_tuple in split_table_set,
+            )
```

Comment:
Let's make this method really clean since we're here. Reading it in English, you have something like:

> `_visualization_hack()` first transforms each entry in the given row, then updates each chart key by getting and setting its value. To transform an entry, it does the following:
>
> * If the entry is a `dict`, it recursively transform each entry in it
> * If the entry is a `Visualize` instance, it invokes the `_config_callback`
> * If the entry is a `CustomChart`, it invokes the `_config_callback` using the result of `get_config_value` and records the path to it as a chart key

I would say this isn't a clear description. There's too much distance between using and discovering chart keys, and to understand the logic with `_split_table` and `get_config_value` you have to read their implementations and the rest of the code. Generally speaking, having to track state (like `chart_keys`) always increases complexity, and it's usually possible to avoid it.

The simplest description I can come up for this method is:

> `_visualization_hack()` processes visualizations in the given tree. It records each `Visualize` value in the run config and replaces it by its `_data`. It records each `CustomChart` value in the run config and replaces it by its `_data`, also rewriting its key to end with `_table`.

Each sentence corresponds to a method or block of code:

```python
def _visualization_hack(self, tree: Dict[str, Any]) -> None:
    self._process_visualizations(tree, key_prefix="")

def _process_visualizations(self, tree: Dict[str, Any], key_prefix: str) -> None:
    delete_keys = []
    for k, v in tree.items():
        formatted_key = f"{key_prefix}.{k}"

        if isinstance(v, dict):
            self._process_visualizations(tree, formatted_key)

        elif isinstance(v, Visualize):
            self._config_update(
                key=v.get_config_key(formatted_key),
                value=v.get_config_value(formatted_key),
            )
            tree[k] = v._data

        elif isinstance(v, CustomChart):
            self._config_update(
                key=v.get_config_key(formatted_key),
                value=v.get_config_value(
                    "Vega2",
                    ...,
                ),
            )
            delete_keys.append(k)
            if v._split_table:
                tree[f"Custom Chart Tables/{k}_table"] = v._data
            else:
                tree[f"{k}_table"] = v._data

    for k in delete_keys:
        del tree[k]
```

I didn't fill out the `...` part because I'm not sure about it. It looks like:

* If `_split_table` is false, we use `user_query(f"{formatted_key}_table")`
* Otherwise, if the chart is 1 level deep, we use `user_query(f"Custom Chart Tables/{formatted_key}_table")`

But for charts deeper than 1 level with `_split_table` set to true, I'm not sure if the code is correct and what the expected output should be.